src/UnionSums.hs

```diff
@@ -30,22 +32,44 @@ unionSumTypes newNameStr typeNames =
     return [DataD [] (mkName newNameStr) [] Nothing newConstructors []]
   where
     mkNewConstructors typeName = do
-      info <- reify typeName
-      case info of
-        (TyConI (DataD _ _ _ _ constructors _)) ->
-          either fail return $ mapM (modConstructor typeName) constructors
-        _ -> fail $ nameBase typeName ++ " must be a sum type"
+      cons <- getConstructors typeName
+      either fail return $ mapM (modConstructor typeName) cons
 
     modConstructor :: Name -> Con -> Either String Con
     modConstructor typeName (NormalC conName args) =
-        (\n -> (NormalC (mkName $ n ++ newNameStr) args)) <$>
+        (\n -> NormalC (mkName n) args) <$>
         note "Constructor name missing type name suffix"
-          (stripNameSuffix typeName conName)
+          (changeSuffix (nameBase typeName) newNameStr (nameBase conName))
     modConstructor _ _ = fail "Unrecognised constructor pattern"
 
-    stripNameSuffix :: Name -> Name -> Maybe String
-    stripNameSuffix suf = stripSuffix (nameBase suf) . nameBase
+mkConverter :: Name -> Name -> Q [Dec]
+mkConverter subName mainName = do
```

Comment:
I'd change `mainName` here to `unionName` to tie it into the `unionSumTypes` helper a bit more.